Boyer-Moore algorithm updates

mmazzei opened the pull request:
This is what I did:
 - added some unit tests
 - ensure every test pass
 - added Boyer-Moore-Horspool algorithm (previously mentioned in the README but don't implemented) with tests.
 - added me in the credits section ;)

This fixes #284.

kelvinlauKL commented on Boyer-Moore/BoyerMoore.swift:
```diff
@@ -6,7 +6,7 @@
 */
 extension String {
-    func indexOf(pattern: String) -> String.Index? {
+    func indexOf(pattern: String, useHorspoolImprovement: Bool = false) -> String.Index? {
```
I think `index(of pattern: String, useHorspoolImprovement: Bool = false) -> Index?` is a better method signature. 

Since Swift 3, many methods switched over from `indexOf` to `index(of:)`. `String.Index` is unnecessary since we're inside a `String` extension and the compiler can infer the type of `Index`.

kelvinlauKL commented on Boyer-Moore/BoyerMoore.swift:
```diff
@@ -53,15 +53,23 @@ extension String {
+                    // If no match, we can only safely skip one character ahead.
+                    i = index(after: i)
+                }
+                else {
```
Our style guidelines prefer to have the `else` statement on the same line as the closing brackets for the `if` statement.

kelvinlauKL commented on Boyer-Moore/README.markdown:
```diff
@@ -158,40 +158,66 @@ Here's an implementation of the Boyer-Moore-Horspool algorithm:
+        // Cache the length of the search pattern because we're going to
+        // use it a few times and it's expensive to calculate.
+        let patternLength = pattern.characters.count
+        assert(patternLength > 0)
```
I feel like we should replace these asserts with `guard patternLength > 0, characters.count < patternLength else { return nil }`. What do you think?

kelvinlauKL commented on Boyer-Moore/README.markdown:
```diff
@@ -158,40 +158,66 @@ Here's an implementation of the Boyer-Moore-Horspool algorithm:
+        // The pattern is scanned right-to-left, so skip ahead in the string by
+        // the length of the pattern. (Minus 1 because startIndex already points
+        // at the first character in the source string.)
+        var i = self.index(startIndex, offsetBy: patternLength - 1)
```
Unnecessary reference to `self`. There are a couple of these.

mmazzei commented:
Thanks for the comments!

I agree with the improvements and implemented the changes.

Also updated the playground and made other minor changes (added tests, renamed `useHorspoolImprovement` to `usingHorspoolImprovement`).

The new Swift 3 style for method signatures feels more natural.

Regards!

kelvinlauKL commented on Boyer-Moore/BoyerMoore.playground/Contents.swift:
```diff
@@ -1,72 +1,91 @@
-      }
+    func index(of pattern: String, usingHorspoolImprovement: Bool = false) -> Index? {
+        // There are no possible match in an empty string
+        guard !isEmpty else { return nil }
```
This feels a bit redundant since the latter `guard` statement covers this case as well.

kelvinlauKL commented on Boyer-Moore/BoyerMoore.swift:
```diff
@@ -6,33 +6,35 @@
-    func indexOf(pattern: String) -> String.Index? {
+    func index(of pattern: String, usingHorspoolImprovement: Bool = false) -> Index? {
+        // There are no possible match in an empty string
+        guard !isEmpty else { return nil }
```
Same as before, the latter `guard` statement should cover this case just as well.

mmazzei commented:
You're right. Redundant validation removed. 👍

kelvinlauKL commented:
Merging this in! Thanks and happy Boxing Day! @mmazzei
